File: day-2/quest-2.js
import fs from "fs";
import readline from "node:readline";

const file = fs.createReadStream('./text.txt', 'utf8');
const lines = readline.createInterface({
    input: file,
    crlfDelay: Infinity,
});

const input = [];
for await (const line of lines) {
    input.push(line.split(" ").map(item => +item));
}

let isAscending = null;

const ascending = (input, matcher) => {
    return matcher < input;
}


const matchesRules = (row) => {
    const previous = row[0];
    const current = row[1];

    if (previous === current) {
        return false;
    }
    if (isAscending === null) {
        isAscending = ascending(current, previous);
    }

    if (isAscending && !ascending(current, previous)) {
        return false
    }
    if (!isAscending && ascending(current, previous)) {
        return false
    }

    return !(Math.abs(current - previous) > 3);
}

const isSafe = (row, failed = false) => {
    console.log(row);

    if (row.length <= 1) return true;
    if (failed && !matchesRules([...row])) {
        return false;
    }

    const isMatching = matchesRules(row);

    if (!isMatching) {
        row = row.filter((item, index) => index !== 1);
        console.log('didnt', row);
        if (!matchesRules(row)) {
            console.log('failed after second chance')
            return false;
        }
    }

    const item = [...row];
    return isSafe(item.slice(1, item.length), !failed ?? matchesRules(row));
}

const safe = [];
for (const row of input) {
    isAscending = null;
    isSafe(row) ? safe.push(row) : console.log('failed', row);
}

console.log(safe.length);

File: day-2/quest-2.py
rows = []

with open('text.txt') as f:
    for line in f.read().splitlines():
        split_lines = line.split(' ')
        rows.append(list(map(int, split_lines)))


def meets_criteria(number_row: list, is_ascending: bool | None = None) -> (bool, bool):
    if is_ascending is None:
        is_ascending = number_row[0] < number_row[1]

    if (is_ascending and number_row[0] > number_row[1]
            or not is_ascending and number_row[0] < number_row[1]):
        return False, is_ascending

    if abs(number_row[0] - number_row[1]) > 3:
        return False, is_ascending

    if number_row[0] == number_row[1]:
        return False, is_ascending

    return True, is_ascending


def is_safe(number_row: list, initial: bool, is_ascending: bool | None = None, failed: bool = False) -> bool:
    if len(number_row) == 1:
        return True

    is_correct, is_ascending = meets_criteria(number_row, is_ascending)

    if failed and not is_correct:
        return False

    if not is_correct:
        failed = True

    temp = number_row.copy()
    temp.pop(0)

    if len(temp) > 2:
        print('first incorrect', temp, number_row)
        is_correct, is_ascending = meets_criteria(temp, is_ascending)

        if not is_correct:
            number_row.pop(1)
            is_valid, is_ascending = meets_criteria(number_row, is_ascending)

            if not is_valid and initial:
                is_ascending = None
            elif not is_valid:
                return False
            temp = number_row

    return is_safe(temp, is_ascending=is_ascending, failed=failed, initial=False)


safes = []
for row in rows:
    result = is_safe(row.copy(), True)
    if result:
        safes.append(row)
    print("is_safe:", result, row)

print(len(safes))
print(safes)
